### PR TITLE
Add support for a postprocess event

### PR DESCRIPTION
**Pull request by dawesc**
Hi there, i've implemented this using the Handler.cs, i'm not 100% convinced since this means lots of little code changes however it does keep it in the same place as the preprocessor which i liked. Please have a look and see if you think it's acceptable.

Many thanks

Christopher

### CODE REVIEW

**Comment by dawesc** [comment]
Ok strange bug with ordering of array output in mono; will fix this first thing tomorrow.

File: Json-Rpc/Handler.cs [hunk 1]
```diff
@@ -534,10 +517,40 @@ private JsonRpcException PreProcess(JsonRequest request, object context)
...
+        private JsonResponse PostProcess(Action<JsonResponse> callback, JsonRequest request, JsonResponse response, object context)
+        {
+            if (externalPostProcessingHandler != null)
+            {
+                try
+                {
+                    JsonRpcException exception = externalPostProcessingHandler(request, response, context);
+                    if (exception != null)
+                    {
+                        System.Diagnostics.StackTrace t = new System.Diagnostics.StackTrace();
```
Comment (Astn) [inline Json-Rpc/Handler.cs]: Did you mean to pass 't' in the JsonResponse?

File: AustinHarris.JsonRpcTestN/Test.cs [hunk 2]
```diff
@@ -1470,7 +1470,279 @@ public void TestSingleResultBatch()
...
+                this.rpc = rpc;
+                this.context = context;
+
+                return null;
+            }
+        }
+
+        [Test()]
+        public void TestPreProcessor()
+        {
```
Comment (Astn) [inline AustinHarris.JsonRpcTestN/Test.cs]: I Love seeing the tests. Nice work!

File: AustinHarris.JsonRpcTestN/service.cs [hunk 3]
```diff
@@ -303,5 +304,41 @@ public void Notify(string message)
...
+        public string TestPostProcessorThrowsJsonRPCException(string inputValue)
+        {
+            throw new JsonRpcException(-27000, "Just some testing", null);
+        }
+
+        [JsonRpcMethod]
+        public string TestPostProcessorThrowsException(string inputValue)
+        {
+            throw new Exception("TestException");
+        }
```
Comment (Astn) [inline AustinHarris.JsonRpcTestN/service.cs]: We might as well add some tests for Context.SetException while were at it to both of these. Similar to:
`[JsonRpcMethod("error4")]
        private string throwsException3(string s)
        {
            JsonRpcContext.SetException(new JsonRpcException(-27000, "This exception was thrown using: JsonRpcContext.SetException()", null));
            return null;
        }`

File: Json-Rpc/Config.cs [hunk 4]
```diff
@@ -29,6 +39,15 @@ public static void SetPreProcessHandler(PreProcessHandler handler)
         }
 
         /// <summary>
+        /// Sets the the PostProcessing Handler on the default session.
+        /// </summary>
+        /// <param name="handler"></param>
+        public static void SetPostProcessHandler(PostProcessHandler handler)
+        {
+            Handler.DefaultHandler.SetPostProcessHandler(handler);
+        }
```
Comment (Astn) [inline Json-Rpc/Config.cs]: I think we also need one of these functions that will set the handler on a specific sessionId.

File: Json-Rpc/Handler.cs [hunk 5]
```diff
@@ -185,12 +186,6 @@ public void UnRegister(string key)
         /// <returns></returns>
         public JsonResponse Handle(JsonRequest Rpc, Object RpcContext = null, Action<JsonResponse> callback = null)
         {
```
Comment (Astn) [inline Json-Rpc/Handler.cs]: I'm a little confused about the changes I'm seeing here. Was
`Action<JsonResponse> callback = null`
from one of your prior commits? 
If so, any chance you could squash your commits so the code is easier to review?
https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/

File: AustinHarris.JsonRpcTestN/Test.cs [hunk 6]
```diff
@@ -1470,7 +1470,279 @@ public void TestSingleResultBatch()
 				JsonRpcProcessor.Process(@"[{""jsonrpc"":""2.0"",""method"":""ReturnsDateTime"",""params"":{},""id"":1}]");
```
Comment (Astn) [inline AustinHarris.JsonRpcTestN/Test.cs]: Overall great stuff so far, I like where your headed. :+1:

**Comment by Astn** [comment]
Weirdness about the array ordering like you said. :confounded:

File: Json-Rpc/Handler.cs (hunk 1 above)
Comment (dawesc) [inline Json-Rpc/Handler.cs]: Hi sorry, no this is some leftover debug where i had issues with some code paths not running as i expected so i was tracing the origin, i'll remove t from here, thanks for catching this.

File: Json-Rpc/Handler.cs (hunk 5 above)
Comment (dawesc) [inline Json-Rpc/Handler.cs]: Yes it was, will squash them last thing today :)

File: AustinHarris.JsonRpcTestN/service.cs (hunk 3 above)
Comment (dawesc) [inline AustinHarris.JsonRpcTestN/service.cs]: great thanks, i've added these

File: Json-Rpc/Config.cs (hunk 4 above)
Comment (dawesc) [inline Json-Rpc/Config.cs]: I've added support for this by making the method Set{Pre/Post}ProcessHandler public instead of internal, i hope that's ok?

**Comment by dawesc** [comment]
Thanks so much for taking the time to review this so quickly, it's much appreciated. Hopefully everything that you pointed to is now correct the thing i'm most unsure on was the session based handler change. I'm afraid testing the JSON so that travis is happy caused some pain so i'm sorry i've added quite a bit of code just to traverse the data to determine if it contains identical data but not necessarily in the same order, we could use a third party library to do this but wasn't sure about including further dependencies into the test project. Thanks again, Christopher.

**Comment by dawesc** [comment]
Hi Austin, many thanks for your support, it's much appreciated!

**Pull request merged**